src/main.rs

```diff
@@ -1,3 +1,176 @@
+        .filter(|&c| c.is_ascii_graphic() || c.is_ascii_whitespace())
+        .collect()
+}
+
+fn send(
+    serial_port: &mut Box<dyn SerialPort>,
+    buf: &[u8],
+) -> Result<(), Box<dyn std::error::Error>> {
+    serial_port.write_all(buf)?;
+    serial_port.flush()?;
+    Ok(())
+}
+
+fn receive(serial_port: &mut Box<dyn SerialPort>) -> Option<String> {
+    let mut buf: Vec<u8> = vec![0; 1000];
+    let bytes_read = serial_port.read(buf.as_mut_slice()).unwrap_or(0);
+    if bytes_read == 0 {
+        return None;
+    }
+    let s = remove_non_printable(&String::from_utf8_lossy(&buf));
+    if s.is_empty() {
+        return None;
+    }
+    Some(s)
+}
+
+fn enter_u_boot(serial_port: &mut Box<dyn SerialPort>) -> String {
+    let mut console_output = String::new();
+    let mut timeout_counter = 0;
+
+    loop {
+        send(serial_port, b"x").expect("Failed to write to serial port");
+
+        if let Some(s) = receive(serial_port) {
+            console_output += s.as_str();
+            timeout_counter = 0;
+        } else {
+            timeout_counter += 1;
+        }
+
+        if console_output.contains("=>") || timeout_counter >= 100 {
+            break;
+        }
+    }
+    send(serial_port, b"\x03").expect("Failed to write to serial port"); // clear prompt
+    console_output
+}
+
+fn run_u_boot_cmd(serial_port: &mut Box<dyn SerialPort>, cmd: &str) -> String {
+    send(serial_port, format!("{cmd}\n").as_bytes()).expect("Failed to write to serial port");
+
+    let mut console_output = String::new();
+    let mut timeout_counter = 0;
+
+    loop {
+        if let Some(s) = receive(serial_port) {
+            console_output += s.as_str();
+            timeout_counter = 0;
+        } else {
+            timeout_counter += 1;
+        }
+
+        if timeout_counter >= 10 {
+            break;
+        }
+    }
+    console_output
+}
+
+fn report_issue(issue: &str, instructions: &str) {
+    println!("! {issue}");
+    println!("-> {instructions}");
+}
+
+fn run_u_boot_check(
+    serial_port: &mut Box<dyn SerialPort>,
+    cmd: &str,
+    pattern: &str,
+    issue: &str,
+    instructions: &str,
+) -> bool {
+    let console_output = run_u_boot_cmd(serial_port, cmd);
+
+    if !console_output.contains(pattern) {
+        report_issue(issue, instructions);
+        return false;
+    }
+    true
+}
+
+fn analyze(serial_port: &mut Box<dyn SerialPort>) {
+    let early_check_info = vec![
+        ("U-Boot SPL", "No U-Boot detected"),
+        ("DRAM:  128 MiB", "Wrong RAM size detected"),
+        (
+            "Net:   eth0: eth@10110000",
+            "Ethernet could not be initialized",
+        ),
+        ("=>", "Could not enter U-Boot shell"),
+    ];
+    let console_output = enter_u_boot(serial_port);
+
+    for (pattern, issue) in early_check_info {
+        if !console_output.contains(pattern) {
+            report_issue(issue, INSTRUCTIONS_LM);
+            return;
+        }
+    }
+
+    let u_boot_check_info = vec![
```

Comment:
unneeded heap allocation
```suggestion
    let u_boot_check_info = [
```